### PR TITLE
WT-10367 Implement file open, read, close, exists, size and lock for GCP file handle

### PR DESCRIPTION
**Pull request by bzm-jas**
Implemented
- gcp_file_open
- gcp_file_close
- gcp_file_exists
- gcp_file_size
- gcp_file_lock
- gcp_file_read

### CODE REVIEW

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 1]
```diff
@@ -245,24 +261,109 @@ static int
...
+    if ((fs->gcp_conn->object_exists(name, exists, size) == 0) && !exists) {
+        std::cerr << "gcp_file_open: object with name " << name << " does not exist in the bucket."
+                  << std::endl;
+        return EINVAL;
+    }
+
+    // Check if there is already an existing file handle open.
+    std::vector<gcp_file_handle *>::iterator fh_iterator =
+      std::find_if(fs->fh_list.begin(), fs->fh_list.end(),
+        [name](gcp_file_handle *fh) { return strcmp(name, fh->name.c_str()) == 0; });
```
Comment (jiechenbo) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:321]: I would use fh->name.compare(name) instead and follow CPP conventions

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 2]
```diff
@@ -245,24 +261,109 @@ static int
...
+
+    if (fh_iterator != fs->fh_list.end()) {
+        (*fh_iterator)->reference_count++;
+        *file_handle_ptr = reinterpret_cast<WT_FILE_HANDLE *>(*fh_iterator);
+
+        return 0;
+    }
+
+    // If file handle does not exist.
```
Comment (jiechenbo) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:330]: Can we elaborate this comment, what happens when the file handle doesn't exist?

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 3]
```diff
@@ -245,24 +261,109 @@ static int
...
+    gcp_file_handle *gcp_fh;
+    try {
+        gcp_fh = new gcp_file_handle;
+    } catch (std::bad_alloc &e) {
+        std::cerr << "gcp_file_open: " << e.what() << std::endl;
+        return ENOMEM;
+    }
+    gcp_fh->file_system = fs;
+    gcp_fh->name = std::string(name);
```
Comment (jiechenbo) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:339]: I don't think this will work since the lifetime of std::string(name) will be destructed after this line.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 4]
```diff
@@ -245,24 +261,109 @@ static int
...
+    gcp_fh->fh.fh_write = nullptr;
+
+    // We require exclusive access to the list of file handles when adding file handles to it. The
+    // lock_guard will be unlocked automatically when the scope is exited.
+    {
+        std::lock_guard<std::mutex> lock(fs->fh_list_mutex);
+        fs->fh_list.push_back(gcp_fh);
+    }
+
+    *file_handle_ptr = file_handle;
```
Comment (jiechenbo) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:366]: We don't need file_handle variable here, since we can replace it with gcp_fh->fh.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 5]
```diff
@@ -292,7 +393,37 @@ gcp_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *from, c
...
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    std::string object_key = gcp_fh->name;
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
+    bool exists;
+    size_t size;
+
+    // Get file size if the object exists.
+    fs->gcp_conn->object_exists(object_key, exists, size);
+
+    sizep = reinterpret_cast<wt_off_t *>(size);
```
Comment (jiechenbo) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:420]: I think you need to do *sizep

**Review by jiechenbo — COMMENTED** [review]
Inital pre-review

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 6]
```diff
@@ -185,9 +189,21 @@ gcp_add_reference(WT_STORAGE_SOURCE *storage_source)
...
 gcp_file_close(WT_FILE_HANDLE *file_handle, WT_SESSION *session)
 {
-    WT_UNUSED(file_handle);
     WT_UNUSED(session);
 
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
+
+    // We require exclusive access to the list of file handles when removing file handles. The
+    // lock_guard will be unlocked automatically once the scope is exited.
```
[the remark this answers is not in the document]
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here:[687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 1 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 3 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here:[687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 5 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 2 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 4 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here:[687beb0](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/687beb0e754c4e60654dfe49bdeda9dc9bf03665)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 7]
```diff
@@ -192,9 +196,26 @@ gcp_add_reference(WT_STORAGE_SOURCE *storage_source)
 static int
 gcp_file_close(WT_FILE_HANDLE *file_handle, WT_SESSION *session)
 {
-    WT_UNUSED(file_handle);
     WT_UNUSED(session);
 
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+
+    // If there are other active instances of the file being open, do not close file handle.
+    if (--gcp_fh->reference_count != 0) {
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:204]: Can we remove the brackets here as this is a single line `if` statement?

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 8]
```diff
@@ -192,9 +196,26 @@ gcp_add_reference(WT_STORAGE_SOURCE *storage_source)
...
+        return 0;
+    }
+
+    // No more active instances of open file, close the file handle.
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:209]: Is there a reason why we're doing this `reinterpret_cast` thing here? Will `gcp_fh->file_system` work here?

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 9]
```diff
@@ -326,6 +455,27 @@ gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
...
 }
 
+static int
+gcp_file_read(
+  WT_FILE_HANDLE *file_handle, WT_SESSION *session, wt_off_t offset, size_t len, void *buf)
+{
+    WT_UNUSED(session);
+
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    std::string object_key = gcp_fh->name;
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:465]: Do we need this `object_key` variable given that it's only used once?

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 10]
```diff
@@ -326,6 +455,27 @@ gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
...
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:466]: Same as above comment regarding the `reinterpret_cast` use.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 11]
```diff
@@ -326,6 +455,27 @@ gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
...
+
+    int ret;
+
+    if ((ret = fs->gcp_conn->read_object(object_key, offset, len, buf)) != 0) {
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:470]: Same as above comment regarding the bracketing.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 12]
```diff
@@ -281,12 +302,84 @@ static int
...
+    }
+    if (!exists) {
+        std::cerr << "gcp_file_open: object named " << name << " does not exist in the bucket."
+                  << std::endl;
+        return EINVAL;
+    }
+
+    // Check if there is already an existing file handle open.
+    auto fh_iterator = std::find_if(fs->fh_list.begin(), fs->fh_list.end(),
+      [name](gcp_file_handle *fh) { return fh->name.compare(name); });
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:337]: I believe `fh->name.compare(name)` will return 0 if the names match, should we change this to return a boolean instead? (`fh->name.compare(name) == 0`)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 13]
```diff
@@ -316,7 +409,43 @@ gcp_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *from, c
...
+}
+
+static int
+gcp_file_size(WT_FILE_HANDLE *file_handle, WT_SESSION *session, wt_off_t *sizep)
+{
+    WT_UNUSED(session);
+
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    std::string object_key = gcp_fh->name;
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:429]: Same as above comment regarding the `reinterpret_cast` use.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 14]
```diff
@@ -316,7 +409,43 @@ gcp_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *from, c
...
+    return 0;
+}
+
+static int
+gcp_file_size(WT_FILE_HANDLE *file_handle, WT_SESSION *session, wt_off_t *sizep)
+{
+    WT_UNUSED(session);
+
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    std::string object_key = gcp_fh->name;
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:428]: Same comment as before, do we need this object_key variable given that it's only used once?

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 15]
```diff
@@ -316,7 +409,43 @@ gcp_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *from, c
...
+    gcp_file_system *fs = reinterpret_cast<gcp_file_handle *>(gcp_fh)->file_system;
+    bool exists;
+    size_t size;
+    int ret;
+    *sizep = 0;
+
+    // Get file size if the object exists.
+    if ((fs->gcp_conn->object_exists(object_key, exists, size))) {
+        std::cerr << "gcp_file_size: object_exists request to google cloud failed." << std::endl;
+
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:438]: Unnecessary whitespace.

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 9 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp:465]: `object_key` was used for better readability, I'm happy to remove it if it's a bit redundant :))

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 7 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: yup, fixed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 8 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: I agree `gcp_fh->file_system` is more suitable here, fixed in: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 9 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 10 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 11 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 12 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 13 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 14 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 15 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: addressed here: [9faf10f](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/9faf10fe9b57e6c7c29f5909f2cb8c35c5a3ec8e)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 16]
```diff
@@ -58,8 +59,11 @@ static int gcp_remove(WT_FILE_SYSTEM *, WT_SESSION *, const char *, uint32_t)
   __attribute__((__unused__));
 static int gcp_rename(WT_FILE_SYSTEM *, WT_SESSION *, const char *, const char *, uint32_t)
   __attribute__((__unused__));
-static int gcp_file_size(WT_FILE_SYSTEM *, WT_SESSION *, const char *, wt_off_t *)
+static int gcp_file_lock(WT_FILE_HANDLE *, WT_SESSION *, bool) __attribute__((__unused__));
+static int gcp_file_size(WT_FILE_HANDLE *, WT_SESSION *, wt_off_t *) __attribute__((__unused__));
```
Comment (mm-ng) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:63]: Just a genuine question, are these `__attribute__((__unused__))` here on purpose?

Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp:63]: yup, they are used to prevent warnings as some functions have been defined but not used, they will be removed altogether as part of [WT-10425](https://jira.mongodb.org/browse/WT-10425)

Comment (mm-ng) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp:63]: Sounds good, thanks!

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 17]
```diff
@@ -326,6 +452,25 @@ gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
...
+
+    gcp_file_handle *gcp_fh = reinterpret_cast<gcp_file_handle *>(file_handle);
+    gcp_file_system *fs = gcp_fh->file_system;
+
+    int ret;
+
+    if ((ret = fs->gcp_conn->read_object(gcp_fh->name, offset, len, buf)) != 0)
+        std::cerr << "gcp_file_read: read attempt failed." << std::endl;
+    else
+        std::cerr << "gcp_file_read: read succeeded." << std::endl;
```
Comment (tammybailey) [inline ext/storage_sources/gcp_store/gcp_storage_source.cpp:469]: Do we want this output? We don't print "success" output elsewhere.

File: test/suite/test_tiered19.py [hunk 18]
```diff
@@ -110,12 +110,26 @@ def test_gcp_filesystem(self):
         non_exist_bucket = "non_exist"
         self.assertRaisesHavingMessage(wiredtiger.WiredTigerError,
             lambda: ss.ss_customize_file_system(
-                session, non_exist_bucket, None, self.get_fs_config(prefix)), err_msg)
-        # Provide a bucket name that exists but we do not have access to.
+                session, non_exist_bucket, self.auth_token, self.get_fs_config(prefix)), err_msg)
+        # Provide a bucket name that exists but we do not have access to.
```
Comment (tammybailey) [inline test/suite/test_tiered19.py:114]: I think there's a trailing whitespace here.

File: test/suite/test_tiered19.py [hunk 19]
```diff
@@ -110,12 +110,26 @@ def test_gcp_filesystem(self):
...
             lambda: ss.ss_customize_file_system(
-                session, no_access_bucket, None, self.get_fs_config(prefix)), err_msg)
+                session, no_access_bucket, self.auth_token, self.get_fs_config(prefix)), err_msg)
+
+        # Test fs_open_file fails when the target file is not in the bucket and does not exist locally.
+        self.assertRaisesHavingMessage(wiredtiger.WiredTigerError,
+            lambda: fs.fs_open_file(
+                session, 'test_put', file_system.open_file_type_data, file_system.open_readonly), err_msg)
+        
+        # We cannot use the file system to create files, it is readonly.
```
Comment (tammybailey) [inline test/suite/test_tiered19.py:125]: Can we rewrite this comment using complete sentences and as a paragraph rather than one statement per line?

**Review by tammybailey — APPROVED** [review]
A few comments but LGTM otherwise!

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 20]
```diff
@@ -281,12 +301,84 @@ static int
...
+    gcp_fh->fh.fh_map_preload = nullptr;
+    gcp_fh->fh.fh_unmap = nullptr;
+    gcp_fh->fh.fh_read = gcp_file_read;
+    gcp_fh->fh.fh_size = gcp_file_size;
+    gcp_fh->fh.fh_sync = nullptr;
+    gcp_fh->fh.fh_sync_nowait = nullptr;
+    gcp_fh->fh.fh_truncate = nullptr;
+    gcp_fh->fh.fh_write = nullptr;
+
+    // Exclusive Access is required when adding file handles to list of file handles.
```
[the remark this answers is not in the document]
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: fixed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 21]
```diff
@@ -316,7 +408,41 @@ gcp_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *from, c
 }
 
 static int
-gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, wt_off_t *sizep)
+gcp_file_lock(WT_FILE_HANDLE *file_handle, WT_SESSION *session, bool lock)
+{
+    // Locks are always granted.
+    WT_UNUSED(file_handle);
```
[the remark this answers is not in the document]
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: changed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp [hunk 22]
```diff
@@ -326,6 +452,25 @@ gcp_file_size(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name
     return 0;
 }
 
+static int
+gcp_file_read(
+  WT_FILE_HANDLE *file_handle, WT_SESSION *session, wt_off_t offset, size_t len, void *buf)
+{
+    WT_UNUSED(session);
```
[the remark this answers is not in the document]
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: changed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

File: ext/storage_sources/gcp_store/gcp_storage_source.cpp (hunk 17 above)
Comment (bzm-jas) [reply ext/storage_sources/gcp_store/gcp_storage_source.cpp]: removed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

File: test/suite/test_tiered19.py (hunk 18 above)
Comment (bzm-jas) [reply test/suite/test_tiered19.py]: fixed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

File: test/suite/test_tiered19.py (hunk 19 above)
Comment (bzm-jas) [reply test/suite/test_tiered19.py]: addressed here: [1487000](https://github.com/wiredtiger/wiredtiger/pull/8777/commits/1487000e645e0a2290646fdcf28759a1dab5ce96)

**Review by mm-ng — APPROVED** [review]
LGTM

**Comment by bzm-jas** [comment]
evergreen merge
